shell.nix

```diff
@@ -1,11 +1,19 @@
-# Copyright (C) 2005 - 2021 Settlers Freaks <sf-team at siedler25.org>
+# Copyright (C) 2005 - 2022 Settlers Freaks <sf-team at siedler25.org>
 #
 # SPDX-License-Identifier: GPL-2.0-or-later
 
-{ pkgs ? import <nixpkgs> {} }:
+{
+  pkgs ? import <nixpkgs> { overlays = [
+    (final: prev: {
+      SDL2 = prev.SDL2.override {
+        withStatic = true;
```

Comment:
What does this actually do? Note that in upstream SDL2 for either autotools as well as the CMake build a static SDL2main library is always build. And we don't need a static SDL2, i.e. a `libSDL2.a`

It appears to me to be rather a bug in the Nix SDL2 build/installation, but would need clarification here. I think an SDL2 installation without the libSDL2main.a library can be considered incomplete. If that isn't the case we could also make the dependency on that library optional in our CMake find module as IIRC we don't need that anyway,

Reply:
Maybe as additional info about what does it do, as i am not sure if you asked why i am doing it, or what is this technically.


```
<package> = prev.<package>.override {
        <some inputs>
};
```
is the nix packages pattern to override a package definition from the repository.
So there is a package SDL in that case, which has some inputs, but in  general everything from the package definition can be overriden. 
The SDL package itself has a input variable, which can be set, to define that static files should be kept in place after compiling, whilst the default behavior deletes them after compile was done.

By the shown definition, its defined, that the shell build by `shell.nix` should make a custom version of SDL on the PATH available, which still has the static files.

What can be set or overriden depends on the packages in the nixpkgs.
For SDL2 that build definion is located here:

https://github.com/NixOS/nixpkgs/blob/3838b7e07e8a79ece6cb249eb4bf0a2bfd6d388a/pkgs/development/libraries/SDL2/default.nix#L48


I hope that answers the technical side of what the heck is done there.
Perma link jumps to the parameter definition.

Reply:
>  But libSDL2main.a is somehow special and needed by cmake to read the versions.

To be exact: The SDL2 installation consists of static or dynamic libraries AND this static library. CMake tries to find a (valid) SDL2 installation and as this static library is part of that the search fails if it is missing. So:

> Nixpkgs removes all .a files when not including withStatic,

Is a bug in Nix which must not remove the (always) static libSDLmain library, see https://github.com/NixOS/nixpkgs/issues/146759#issuecomment-974718200

I'm fine with including the workaround here if that does work but please add a comment above it to clearly mark it as a workaround for a Nix bug and possibly even link the issue there so it can be removed once the Nix package is fixed. Maybe this explanation is also useful for the Nix issue, although I think there should be a separate one about the missing `libSDL2main.a` not generically `*.a`. I'll leave that to you as you're more familiar with the Nix stuff.

Or correct me if I misunderstood anything.

Reply:
Absolut correct summarized. Ill add the mentioned improvement, as i am watcher on the PR which should fix  the libSDLmain.a topic ill get a notification if that gets merged downstream, in the meantime i am also trying to get some attention to that on their communication platforms, as the package maintainer on their side seems not very active (but thats just guessing on public github activity)